libraries/botframework-schema/src/ActivityEx.ts

```diff
@@ -0,0 +1,342 @@
+        return isActivity(source, ActivityTypes.Handoff) ? this : null;
+    }
+
+    export function HasContent(source: Partial<Activity>): boolean  {
+        if(source.text) {
+            return true;
+        }
+
+        if(source.summary) {
+            return true;
+        }
+
+        if(source.attachments) {
+            return true;
+        }
+
+        if(source.channelData) {
+            return true;
+        }
+
+        return false;
+    }
+
+    export function GetMentions(source: Partial<Activity>) : Mention[] {
+        return source.entities.filter(x => x.type.toLowerCase() === 'mention')
+        .map(e => e as Mention );
+    }
+
+    export function GetConversationReference(source: Partial<Activity>) : Partial<ConversationReference> {
+        return {
+            activityId: source.id,
+            user: source.from,
+            bot: source.recipient,
+            conversation: source.conversation,
+            channelId: source.channelId,
+            locale: source.locale,
+            serviceUrl: source.serviceUrl
+        } as ConversationReference;
+    }
+
+    export function GetReplyConversationReference(source: Partial<Activity>, reply: Partial<ResourceResponse>) : Partial<ConversationReference> {
+        let reference = GetConversationReference(source);
+        reference.activityId = reply.id;
+        
+        return reference;
+    }
+
+    export function applyConversationReference(source: Partial<Activity>, reference: Partial<ConversationReference>, isComming?: boolean): Partial<Activity> {
+        if (reference.channelId !== undefined) {
+            source.channelId = reference.channelId;
+        }
+
+        if (reference.serviceUrl !== undefined) {
+            source.serviceUrl = reference.serviceUrl;
+        }
+        
+        if (reference.conversation !== undefined) {
+            source.conversation = reference.conversation;
+        }
+
+        if (reference.locale !== undefined) {
+            source.locale = reference.locale;
+        }
+
+        if(isComming) {
+            if (reference.user !== undefined) {
+                source.from = reference.user;
+            }
+
+            if (reference.bot !== undefined) {
+                source.recipient = reference.bot;
+            }
+
+            if(reference.activityId !== undefined) {
+                source.id = reference.activityId;
+            }
+        } else {
+            if(reference.bot !== undefined) {
+                source.from = reference.bot;
+            }
+            if(reference.user !== undefined) {
+                source.recipient = reference.user;
+            }
+            if(reference.activityId !== undefined) {
+                source.replyToId = reference.activityId;
+            }
+        }
+
+        return source;
+    }
+
+    export function IsFromStreamingConnection(source: Partial<Activity>) : boolean {
+        let isHttp = source.serviceUrl.toLowerCase().startsWith("http");

```

Comment:
The variable could be a `const`, as its entire value doesn't change.

Reply:
Feedback applied. Thanks!